tokio/src/sync/batch_semaphore.rs

```diff
@@ -0,0 +1,727 @@
+                (next, needed)
+            } else {
+                remaining = (needed - acquired) - curr;
+                (0, curr)
+            };
+
+            if remaining > 0 && lock.is_none() {
+                // No permits were immediately available, so this permit will
+                // (probably) need to wait. We'll need to acquire a lock on the
+                // wait queue.
+                lock = Some(self.waiters.lock().unwrap());
+                // While we were waiting to lock the wait list, additional
+                // permits may have been released. Therefore, we will acquire a
+                // new snapshot of the current state of the semaphore before
+                // actually enqueueing the waiter..
+                continue;
+            }
+
+            match self.permits.compare_exchange_weak(
+                curr,
+                next,
+                AcqRel,
+                Acquire,
+            ) {
+                Ok(_) => {
+                    acquired += acq;
+                    if remaining == 0 {
+                        return Ready(Ok(()));
+                    }
+                    break lock.unwrap();
+                }
+                Err(actual) => curr = actual,
+            }
+            drop(lock.take());
+        };
+
+        if waiters.closed {
+            return Ready(Err(AcquireError(())));
+        }
+
+        let next = match acquired.cmp(&state) {
+            // if the waiter is in the unqueued state, we need all the requested
+            // permits, minus the amount we just acquired.
+            _ if state >= Waiter::UNQUEUED => needed as usize - acquired,
+            // We have acquired all the needed permits!
+            cmp::Ordering::Equal => 0,
+            cmp::Ordering::Less => state - acquired,
+            cmp::Ordering::Greater => {
+                unreachable!("cannot have been assigned more than needed permits")
+            }
+        };
+        
+        // Typically, one would probably expect to see a compare-and-swap loop
+        // here. However, in this case, we don't need to loop. Our most snapshot
+        // of the node's current state was acquired after we acquired the lock
+        // on the wait list. Because releasing permits tPPo waiters also
+        // requires locking the wait list, the state cannot have changed since
+        // we acquired this snapshot.
+        node.state
+            .compare_exchange(state, next, AcqRel, Acquire)
+            .expect("state cannot have changed while the wait list was locked");
+
+        if next == 0 {
+            return Ready(Ok(()));
+        }
+
+        // otherwise, register the waker & enqueue the node.
+        node.waker.with_mut(|waker| 
+            // safety: the wait list is locked, so we may modify the waker.
+            unsafe { *waker = Some(cx.waker().clone()) }
+        );
+
+        unsafe {
+            // XXX(eliza) T_T
```

Comment:
What does this mean?

Reply:
I was hoping it would be possible to implement `Linked` with `Pin`s as the handle type, so that the 
```rust

            let node = Pin::into_inner_unchecked(node) as *mut _;
            let node = NonNull::new_unchecked(node);
```
could be removed, as it felt pretty unfortunate. Sadly, I don't think that's possible with the current linked list APIs. I can remove the comment.